### PR TITLE
Let A Node Disseminate Itself After Join

### PR DESCRIPTION
**Pull request by CorgiMan**
If a node get's a join request, it no longer adds the joining node to the membership.
A node is therefor expected to disseminate it's own existance to the membership.

T212923

### CODE REVIEW

File: test/integration/admin_test.js [hunk 1]
```diff
@@ -24,7 +24,8 @@ var RingpopClient = require('../../client.js');
 var testRingpopCluster = require('../lib/test-ringpop-cluster.js');
 
 testRingpopCluster({
-    size: 1
+    size: 1,
+    dontWaitForConvergence: true,
```
Comment (jwolski) [inline test/integration/admin_test.js]: can you flip this property around and make it read `waitForConvergence`? The double negatives "dont" and "false" make it confusing.

File: test/integration/gossip_test.js [hunk 2]
```diff
@@ -157,8 +158,10 @@ testRingpopCluster({
     size: 5,
     tap: function tap(cluster) {
         mkBadPingReqResponder(cluster[3]);
+    },
+    tapAfterConvergence: function tapAfterConvergence(cluster) {
         mkNoGossip(cluster);
-    }
+    },
```
Comment (jwolski) [inline test/integration/gossip_test.js]: trailing comma; get rid of it.

File: test/integration/gossip_test.js [hunk 3]
```diff
@@ -181,9 +184,9 @@ testRingpopCluster({
 
 testRingpopCluster({
     size: 5,
-    tap: function tap(cluster) {
+    tapAfterConvergence: function tapAfterConvergence(cluster) {
```
Comment (jwolski) [inline test/integration/gossip_test.js]: trailing comma

File: test/integration/join-test.js [hunk 4]
```diff
@@ -92,7 +93,8 @@ testRingpopCluster({
     tap: function tap(cluster) {
         cluster[1].denyJoins();
         cluster[2].denyJoins();
-    }
+    },
+    dontWaitForConvergence: true,
```
Comment (jwolski) [inline test/integration/join-test.js]: trailing comma

File: test/integration/join-test.js [hunk 5]
```diff
@@ -107,15 +109,22 @@ testRingpopCluster({
 });
 
 testRingpopCluster({
-    size: 25
+    size: 25,
```
Comment (jwolski) [inline test/integration/join-test.js]: trailing comma

File: test/integration/join-test.js [hunk 6]
```diff
@@ -107,15 +109,22 @@ testRingpopCluster({
...
-    assert.equal(cluster.length, 25, 'cluser of 25');
-
-    cluster.forEach(function eachNode(node) {
-        assert.ok(node.isReady, 'node is bootstrapped');
+    cluster.forEach(function eachRingpop(ringpop) {
+            ringpop.gossip.stop();
     });
 
-    assert.end();
+    // wait 50 milliseconds to let pings die out, otherwise tchannel error
```
Comment (jwolski) [inline test/integration/join-test.js]: i dont think we should add any explicit waits in our test suite, if we dont have to. why isn't this satisfied by the "wait for convergence"?

File: test/lib/test-ringpop-cluster.js [hunk 7]
```diff
@@ -127,12 +129,53 @@ function testRingpopCluster(opts, name, test) {
     }
 
     tape(name, function onTest(assert) {
-        var cluster = bootstrapClusterOf(opts, function onBootstrap(results) {
-            assert.on('end', function onEnd() {
-                destroyCluster(cluster);
+        var cluster = createClusterOf(opts);;
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: double semi-colons

File: test/lib/test-ringpop-cluster.js [hunk 8]
```diff
@@ -67,6 +66,8 @@ function bootstrapClusterOf(opts, onBootstrap) {
             }, cb);
         });
         ringpop.__channel.listen(Number(parts[1]), parts[0]);
+
+        // ringpop.dissemination.on('changesExhausted', bootstrapped);
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: no need to keep this here.

File: test/lib/test-ringpop-cluster.js [hunk 9]
```diff
@@ -127,12 +129,53 @@ function testRingpopCluster(opts, name, test) {
...
+
+        var nodeExhausted = _.after(cluster.length, onConverged);
+        
+        cluster.forEach(function each(ringpop, i) {
+            ringpop.dissemination.on('changesExhausted', nodeExhausted);
+        });
+
+        cluster = bootstrapClusterOf(cluster, opts, function onBootstrap(results) {
+            if (typeof opts.tapAfterConvergence === 'function') {
+                opts.tapAfterConvergence(cluster);
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: is this actually being called "after convergence"?

File: test/lib/test-ringpop-cluster.js [hunk 10]
```diff
@@ -127,12 +129,53 @@ function testRingpopCluster(opts, name, test) {
...
 
     tape(name, function onTest(assert) {
-        var cluster = bootstrapClusterOf(opts, function onBootstrap(results) {
-            assert.on('end', function onEnd() {
-                destroyCluster(cluster);
+        var cluster = createClusterOf(opts);;
+        var joinResults;
+        var minProtocolPeriod = opts.minProtocolPeriod;
+        opts.minProtocolPeriod = 1;
+
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: remove double blank lines (only one is fine)

File: test/lib/test-ringpop-cluster.js [hunk 11]
```diff
@@ -127,12 +129,53 @@ function testRingpopCluster(opts, name, test) {
     }
 
     tape(name, function onTest(assert) {
-        var cluster = bootstrapClusterOf(opts, function onBootstrap(results) {
-            assert.on('end', function onEnd() {
-                destroyCluster(cluster);
+        var cluster = createClusterOf(opts);;
+        var joinResults;
+        var minProtocolPeriod = opts.minProtocolPeriod;
+        opts.minProtocolPeriod = 1;
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: it's best not to mutate the opts being passed in by the caller

File: test/lib/test-ringpop-cluster.js [hunk 12]
```diff
@@ -127,12 +129,53 @@ function testRingpopCluster(opts, name, test) {
...
+
+
+        function onConverged() {
+            // do not run onConverged if not all joins are succesful
+            if (joinResults === undefined) {
+                return;
+            }
+
+            if (!(opts.dontWaitForConvergence === true || opts.checkChecksums === false)) {
+                var checksums = _.uniq(_.pluck(_.pluck(cluster, 'membership'), 'checksum'))
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: prefer underscore chaining  to nested underscore calls. something like:

```
var checksums = _.chain(cluster)
    .pluck('membership')
    .pluck('checksum');
    .uniq()
    .value();
```

File: test/lib/test-ringpop-cluster.js [hunk 13]
```diff
@@ -127,12 +129,53 @@ function testRingpopCluster(opts, name, test) {
...
+                if (!(checksums.length === 1 && typeof checksums[0] === 'number')) {
+                    assert.fail('not all checksums are equal')
+                }
+            }
+
+            cluster.forEach(function each(ringpop, i) {
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: `i` is unused.

File: test/integration/admin_test.js (hunk 1 above)
Comment (CorgiMan) [inline test/integration/admin_test.js]: I found this a little bit more understandable. This will never be set to false since the default is to wait for convergence. But it's no problem to change.

File: test/integration/join-test.js (hunk 4 above)
Comment (CorgiMan) [inline test/integration/join-test.js]: Yeah.. that's a go thing I guess :)

File: test/integration/join-test.js (hunk 6 above)
Comment (CorgiMan) [inline test/integration/join-test.js]: Since the protocolperiod is set to 1ms, a lot of pings are flying around. After convergence the protocol period is set back to the default but it takes a moment for all the 1ms pings to be handled. If we destroy the channels at this moment, a tchannel error will be generated. This was the quick fix.

File: test/lib/test-ringpop-cluster.js (hunk 9 above)
Comment (CorgiMan) [inline test/lib/test-ringpop-cluster.js]: Nope, I have corrected it.

What do the mkNoGossip functions do exactly. The tests seem to pass wether they're on or commented out.

File: test/integration/join-test.js (hunk 6 above)
Comment (jwolski) [inline test/integration/join-test.js]: Ok, let's assess how long the non-quick fix will take.

File: test/integration/gossip_test.js (hunk 3 above)
Comment (CorgiMan) [inline test/integration/gossip_test.js]: fixed

File: test/integration/join-test.js (hunk 6 above)
Comment (CorgiMan) [inline test/integration/join-test.js]: Took a look at it. The channel.close method has a callback that is currently not used. So I think the function returns with the channel being partly closed.

File: lib/gossip/dissemination.js [hunk 14]
```diff
@@ -156,6 +156,9 @@ Dissemination.prototype._issueAs = function _issueAs(filterChange, mapChanges) {
 
         if (change.piggybackCount > this.maxPiggybackCount) {
             delete this.changes[address];
+            if (Object.keys(this.changes).length === 0) {
```
Comment (jwolski) [inline lib/gossip/dissemination.js]: can you add a `lib/gossip/events.js` file and define a ChangesExhaustedEvent therein (take a look at lib/ring/events.js) for an example. Recently, I've been trying to be a little bit more explicit about the events that Ringpop emits in its various modules.

then this file'd look like:

```
this.emit('changesExhaused', new events.ChangesExhaustedEvent());
```

File: test/integration/gossip_test.js [hunk 15]
```diff
@@ -59,10 +59,11 @@ function mkBadPingReqResponder(ringpop) {
 }
 
 testRingpopCluster({
-    tap: function tap(cluster) {
-        mkNoGossip(cluster);
+    tapAfterConvergence: function tapAfterConvergence(cluster) {
```
Comment (jwolski) [inline test/integration/gossip_test.js]: if not needed, can you remove altogether?

File: test/integration/gossip_test.js [hunk 16]
```diff
@@ -59,10 +59,11 @@ function mkBadPingReqResponder(ringpop) {
...
+        // mkNoGossip(cluster);
     }
 }, 'ping-reqs 1 member', function t(bootRes, cluster, assert) {
+
```
Comment (jwolski) [inline test/integration/gossip_test.js]: take out blank line.

File: test/integration/gossip_test.js [hunk 17]
```diff
@@ -155,8 +156,11 @@ testRingpopCluster({
 
 testRingpopCluster({
     size: 5,
+    // minProtocolPeriod: 1,
```
Comment (jwolski) [inline test/integration/gossip_test.js]: remove

File: test/lib/test-ringpop-cluster.js [hunk 18]
```diff
@@ -127,12 +126,66 @@ function testRingpopCluster(opts, name, test) {
...
+                var checksums = _.chain(cluster)
+                    .pluck('membership')
+                    .pluck('checksum')
+                    .uniq();
+                if (!(checksums.length === 1 && typeof checksums[0] === 'number')) {
+                    assert.fail('not all checksums are equal');
+                }
+            }
+
+            // cluster.forEach(function each(ringpop) {
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: remove

File: test/lib/test-ringpop-cluster.js [hunk 19]
```diff
@@ -127,12 +126,66 @@ function testRingpopCluster(opts, name, test) {
...
+            //     ringpop.minProtocolPeriod = minProtocolPeriod;
+            // });
+
+            test(joinResults, cluster, assert);
+        }
+
+        var nodeExhausted = _.after(cluster.length, onConverged);
+        
+        cluster.forEach(function each(ringpop) {
+            ringpop.dissemination.on('changesExhausted', nodeExhausted);
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: can you rename event handler to `onExhausted`? That's a convention that screams out "callback"

File: test/lib/test-ringpop-cluster.js [hunk 20]
```diff
@@ -127,12 +126,66 @@ function testRingpopCluster(opts, name, test) {
...
     tape(name, function onTest(assert) {
-        var cluster = bootstrapClusterOf(opts, function onBootstrap(results) {
-            assert.on('end', function onEnd() {
-                destroyCluster(cluster);
+        var tmpMinProtocolPeriods = [];
+        var cluster = createClusterOf(opts);
+        var joinResults;
+
+        // store protocol period. Set protocol period to 1 so that convergence is fast
+        cluster.forEach(function each(ringpop, i) {
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: instead of doing this, wouldnt it be easier if in createClusterOf you had:

```
createRingpop(_.extend({
    host: '127.0.0.1',
    port: 10000 + i,
    minProtocolPeriod: 1
}, opts)));
```

File: test/lib/test-ringpop-cluster.js [hunk 21]
```diff
@@ -127,12 +126,66 @@ function testRingpopCluster(opts, name, test) {
...
+        });
+
+        function onConverged() {
+            // do not run onConverged if not all joins are succesful
+            if (joinResults === undefined) {
+                return;
+            }
+
+            // revert back to original protocol period
+            cluster.forEach(function each(ringpop, i) {
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: hm, I see what you're doing here. Is it necessary to revert?

File: test/lib/test-ringpop-cluster.js [hunk 22]
```diff
@@ -127,12 +126,66 @@ function testRingpopCluster(opts, name, test) {
...
+                ringpop.gossip.minProtocolPeriod = tmpMinProtocolPeriods[i];
             });
 
-            test(results, cluster, assert);
+            if (typeof opts.tapAfterConvergence === 'function') {
+                opts.tapAfterConvergence(cluster);
+            }
+
+            if (opts.waitForConvergence !== false && opts.checkChecksums === true) {
```
Comment (jwolski) [inline test/lib/test-ringpop-cluster.js]: this is getting a bit messy. can you extract these 10-12 lines of code into a function like:

```
function assertEqualChecksums(...) {
  ...
}
```

File: test/integration/gossip_test.js (hunk 15 above)
Comment (CorgiMan) [inline test/integration/gossip_test.js]: Why was this needed in the first place? We need gossip for the ring to converge. We could turn gossip off after convergence but I'm not sure why

File: test/lib/test-ringpop-cluster.js (hunk 21 above)
Comment (CorgiMan) [inline test/lib/test-ringpop-cluster.js]: If we for example give minProtocolPeriod to opts we want to quickly converge with minProtocolPeriod of 1 and then revert it to what opts.minProtocolPeriod is. That was the idea, I don't think it is used though

File: test/integration/gossip_test.js (hunk 17 above)
Comment (CorgiMan) [inline test/integration/gossip_test.js]: removed

File: lib/gossip/dissemination.js (hunk 14 above)
Comment (CorgiMan) [inline lib/gossip/dissemination.js]: check

File: test/integration/join-test.js [hunk 23]
```diff
@@ -106,17 +107,20 @@ testRingpopCluster({
...
 
-testRingpopCluster({
-    size: 25
-}, 'mega cluster', function t(bootRes, cluster, assert) {
-    assert.equal(cluster.length, 25, 'cluser of 25');
 
-    cluster.forEach(function eachNode(node) {
-        assert.ok(node.isReady, 'node is bootstrapped');
-    });
+
```
Comment (CorgiMan) [inline test/integration/join-test.js]: Will add this again...

File: test/lib/test-ringpop-cluster.js (hunk 22 above)
Comment (CorgiMan) [inline test/lib/test-ringpop-cluster.js]: check
